Add Fluxes for Generalized Harmonic

osheamonn opened the pull request:
## Proposed changes

Adds normal dot fluxes for GH

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

osheamonn committed:
Add Fluxes for Generalized Harmonic

mar-celine commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -80,4 +81,56 @@ struct ComputeDuDt {
+ * \f}
+ *
+ * where \f$\psi_{ab}\f$ is the spacetime metric, \f$\Pi_{ab}\f$ its conjugate
+ * momentum, \f$ \Phi_{iab} \f$ is the auxiliary field defined as in Phi,
```
I am confused by "defined as in Phi" - is this referring to the code? I would use `'phi'` to make it more clear

osheamonn replied:
I'm referencing the Tag `Phi` which defined what Phi is relative to the spacetime metric. The goal here was so that it links to Phi in the rendered docs, but you're right it would be clearer if I said `Tags::Phi` or something.

mar-celine replied:
Ah, ok that makes more sense :+1:

mar-celine commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -248,9 +249,71 @@ void ComputeDuDt<Dim>::apply(
+    const tnsr::i<DataVector, Dim>& unit_normal) noexcept {
+  const auto shift_dot_normal = get(dot_product(shift, unit_normal));
+
+  auto upper_normal_dot_phi =
```
I missed what the meaning of "upper" was?

osheamonn replied:
The normal is a lower index tensor, so this is the "upper normal" in that the index is raised. I can just change it to normal dot phi since there's only one way that can make sense.

mar-celine commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -80,4 +81,56 @@ struct ComputeDuDt {
+ * \details The expressions for the fluxes is obtained
+ * from <a href="https://arxiv.org/abs/gr-qc/0512093">
+ * gr-qc/0512093 </a>.
+ * The fluxes for each variable are given by taking the principal part of
```
given -> obtained

mar-celine commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -80,4 +81,56 @@ struct ComputeDuDt {
+ * \details The expressions for the fluxes is obtained
+ * from <a href="https://arxiv.org/abs/gr-qc/0512093">
+ * gr-qc/0512093 </a>.
+ * The fluxes for each variable are given by taking the principal part of
```
I don't know what "principal part" means - according to the paper, these are the terms carrying the derivatives of the fields? Not asking you to change it just asking for my own clarification

osheamonn replied:
Yes, the parts of the equations that contains spatial derivatives of the fields. So the flux for each variable is determined solely by terms containing derivatives, the source terms don't contribute.

kidder commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp:
```diff
@@ -80,4 +81,56 @@ struct ComputeDuDt {
+template <size_t Dim>
+struct ComputeNormalDotFluxes {
+ public:
+  using return_tags =
```
don't need return_tags

osheamonn committed:
Exclude ScalarWave Equations implementations from dox

I was getting dox errors that the implementation of ComputeNormalDotFluxes for Scalar Wave was trying to match to the definitions for both ScalarWave and GeneralizedHarmonic. Excluding the implementation from the docs fixes this.

osheamonn committed:
Fix IWYU for ScalarWave Equations

kidder commented:
@nilsdeppe please look at
